data_node/upload_manager.go

```diff
@@ -0,0 +1,123 @@
+package datanode
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"path"
+	"strconv"
+
+	"github.com/SayedAlesawy/Videra-Ingestion/orchestrator/utils/errors"
+	"github.com/julienschmidt/httprouter"
+)
+
+// NewUploadManager A function to create a new upload manager
+func NewUploadManager() UploadManager {
+	return UploadManager{
+		fileBase:  make(map[string]FileInfo),
+		logPrefix: "[Upload-Manager]",
+	}
+}
+
+// HandleUpload is upload endpoint handler
+func (um *UploadManager) HandleUpload(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	log.Println(um.logPrefix, fmt.Sprintf("Received request from %s", r.RemoteAddr))
+	reqType := r.Header.Get("Request-Type")
+
+	switch reqType {
+	case "INIT":
+		um.handleInitialUpload(w, r)
+	case "APPEND":
+
+	default:
+		log.Println(um.logPrefix, r.RemoteAddr, fmt.Sprintf("request-type header value undefined - %s", reqType))
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("Request-Type header value is not undefined"))
+	}
+}
+
+func (um *UploadManager) handleInitialUpload(w http.ResponseWriter, r *http.Request) {
+	log.Println(um.logPrefix, r.RemoteAddr, "Received INIT request")
+
+	err := um.checkInitialUploadHeaders(&r.Header, false)
+
+	if err != nil {
+		log.Println(um.logPrefix, r.RemoteAddr, err)
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(err.Error()))
+		return
+	}
+
+	filesize, _ := strconv.ParseInt(r.Header.Get("Content-Length"), 10, 64)
+	id := generateRandomString(10)
+	filepath := id
+	filename := r.Header.Get("Filename") // Maybe be changed later
+
+	log.Println(um.logPrefix, r.RemoteAddr, "creating file with id", id)
+	err = createFileDirectory(filepath, 0744)
+	if errors.IsError(err) {
+		log.Println(um.logPrefix, r.RemoteAddr, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("Internal server error"))
+		return
+	}
+
+	err = um.addNewFile(id, filepath, filename, filesize)
+	if errors.IsError(err) {
+		log.Println(um.logPrefix, r.RemoteAddr, err)
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("Internal server error"))
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	w.Header().Set("ID", id)
+}
+
+func (um *UploadManager) checkInitialUploadHeaders(h *http.Header, checkToken bool) error {
+	if checkToken {
+		token := h.Get("token")
+
+		if len(token) == 0 {
+			return errors.New("No token provided")
+		}
```

Comment:
Don't we need to make sure that it matches the token granted by the Name Node? We can skip this now, but just a note.